assembly/README.md

```diff
@@ -21,15 +21,75 @@ let assembler = Assembler::default();
 let program = assembler.compile("begin push.3 push.5 add end").unwrap();
 ```
 
+## Assembler options
+By default, the assembler is instantiated in the most minimal form. To extended the capabilities of the assembler, you can apply a chain of `with_*` methods to the default instance in a builder pattern. The set of currently available options is described below.
+
+### Module provider
+To enable calls to procedures from external modules, the assembler must be supplied with a `ModuleProvider`. A module provide tells the assembler how to resolve calls to external procedures.
+
+`ModuleProvider` is a trait which can be implemented in a number of different ways. We have implemented it for the Miden [standard library](../stdlib). Thus, for example, to make Miden stdlib available to programs during compilation, the assembler can be instantiated as follows:
+
+```Rust
+use miden_assembly::Assembler;
+use miden_stdlib::StdLibrary;
+
+// instantiate the assembler with access to Miden stdlib
+let assembler = Assembler::default().with_module_provider(StdLibrary::default());
+```
+Programs compiled with this assembler will be able to invoke any procedures from Miden `stdlib`. For example, something like this will be possible:
+```
+use.std::math::u64
+
+begin
+    push.1.0
+    push.2.0
+    exec.u64::checked_add
+end
+```
+
+### Program kernels
+A *program kernel* defines a set of procedures which can be invoked via `syscall` instructions. Miden programs are always compiled against some kernel, and by default this kernel is empty (i.e., no `syscall`'s are possible).
+
+Instantiating the assembler with a non-empty kernel can be done like so:
+```Rust
+use miden_assembly::Assembler;
+
+// source code of the kernel module
+let kernel_source = "export.foo add end";
+
+// instantiate the assembler with a kernel
+let assembler = Assembler::default().with_kernel(kernel_source);
+```
+
+Programs compiled with this assembler will be able to make calls to `foo` procedure by executing `syscall.foo` instruction.
+
```

Comment:
The example here will need to be fairly involved. We could put a simplistic one (making a `syscall` from the root context), but that example would not be something that anyone would ever want to do, and we might disallow things like this in the future.

A more correct example would need to first go into the user context via a `call` instruction, and then make a `syscall` from there - but I'm trying to avoid adding all this extra complexity here.

Reply:
In that case, maybe we should add more involved examples of execution contexts (and all the rest of this) to the examples repo. Once they exist, we could link the example here.

Reply:
Agreed!